### ISSUE TITLE
Clarify antiforgery protection behavior around multiple tabs being open

### ISSUE THREAD

**Issue by nbellowe** [dotnet/AspNetCore.Docs#25564]
Initially I used the "Synchronizer Token Pattern (STP)" pattern described in the docs under "Antiforgery in ASP.NET Core". This led to issues for any users that use my website from multiple tabs, due to STP enforcing page load order (only POST made by the most recently loaded page would work properly).

I think the documentation could be improved by explaining more about this behavior, and recommending that in cases where you don't want to enforce page load ordering, the cookie to header pattern, (also documented but not named) or another CSRF protection pattern, may be suitable. I am happy to make this change myself.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 47b14f57-82ac-a2e2-cbc7-22a81a60f4ac
* Version Independent ID: bffca13c-223f-c61f-9cb2-9da8811eecfa
* Content: [Prevent Cross-Site Request Forgery (XSRF/CSRF) attacks in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-6.0)
* Content Source: [aspnetcore/security/anti-request-forgery.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/anti-request-forgery.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 350915](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350915)

**Comment by nbellowe** [comment]
(on this page) https://docs.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-6.0#antiforgery-in-aspnet-core

**Comment by Rick-Anderson** [comment]
@blowdart  please review.

**Comment by blowdart** [comment]
Sure give it a shot, but it boils down to "We don't support multiple tabs logged in with different users, or one logged in and one anonymous". If you want to discuss other patterns you'll need to discuss the risks in detail.

**Comment by Rick-Anderson** [comment]
@nbellowe  how about in this doc we add an H2, `## Multiple browser tabs`, and explain that's not supported and why it's problematic.

Then create a [new discussion](https://github.com/dotnet/AspNetCore.Docs/discussions) with your mitigation. We can then point to the discussion.

**Comment by nbellowe** [comment]
Sure, my mitigation was essentially another pattern listed on the doc (copied below). I am happy to update the docs, it sounds like we have a consensus. I will make an MR myself, if thats ok. 


```
// https://developer.mozilla.org/en-US/docs/web/api/document/cookie
const xsrfToken = document.cookie
    .split("; ")
    .find(row => row.startsWith("XSRF-TOKEN="))
    .split("=")[1];

const response = await fetch("/JavaScript/FetchEndpoint", {
    method: "POST",
    headers: { "X-XSRF-TOKEN": xsrfToken }
});

if (response.ok) {
    resultElement.innerText = await response.text();
} else {
    resultElement.innerText = `Request Failed: ${response.status}`
}
```

**Comment by nbellowe** [comment]
https://github.com/dotnet/AspNetCore.Docs/pull/25585

**Comment by nbellowe** [comment]
Perhaps I should point out the cookie to header approach that is mentioned later in the document. Thoughts @blowdart would you mind reviewing?

**Comment by aomader** [comment]
Is this documentation actually still correct? @Rick-Anderson 

I just did a short test based on the default `webapp` template with .NET 8 (`dotnet new webapp -o RazorPagesMovie`) and it is very much possible to open multiple tabs and operate forms in any order.

I am really curious now what the actual behavior is.

**Comment by Rick-Anderson** [comment]
Per the doc ` Using multiple tabs can be problematic.`

**Comment by aomader** [comment]
Yeah, but then [two examples](https://learn.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-8.0#multiple-browser-tabs-and-the-synchronizer-token-pattern) are given, which I am unable to reproduce (e.g., "Requests made from previously loaded tabs fail with an error: ...").
So I am wondering what is incorrect: the documentation or the implementation?

**Comment by Tobikblom** [comment]
@Rick-Anderson 

I believe that @aomader is correct about the documentation being incorrect.
 
Per @blowdart's comment:
> [...] it boils down to "We don't support multiple tabs logged in with different users, or one logged in and one anonymous".

The documentation does not reflect this, but instead gives the impression that the user is unable to use multiple tabs in general, which, in my experience, is not the case.

Related issue: #26422 
Related answer on StakOverflow: https://stackoverflow.com/a/78615290

**Comment by aomader** [comment]
@Rick-Anderson I do not see how your adjustment to the docs answer the [initial question](https://github.com/dotnet/AspNetCore.Docs/issues/25564#issuecomment-2262152385). It is still not clear **how to reproduce the two given example cases for when the mentioned issue is supposed to arise.**

**Comment by Rick-Anderson** [comment]
> Yeah, but then [two examples](https://learn.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-8.0#multiple-browser-tabs-and-the-synchronizer-token-pattern) are given, which I am unable to reproduce (e.g., "Requests made from previously loaded tabs fail with an error: ..."). So I am wondering what is incorrect: the documentation or the implementation?

How does it not answer that? Have you thoroughly tested every browser and every version?

**Comment by aomader** [comment]
@Rick-Anderson Excuse me, but that answer is also not helpful. Instead of me doing an exhaustive trial and error search across all kinds of browsers, browser settings and devices, it would be much more helpful if you instead explicitly tell us what the necessary conditions are that would cause the aforementioned issue. How am I otherwise supposed to judge the impact this protection is having?

**Comment by Tobikblom** [comment]
@blowdart's pevious comment (https://github.com/dotnet/AspNetCore.Docs/issues/25564#issuecomment-1093051421) provided the answer to how antiforgery tokens behave. 

This is what the the documentation should reflect *if* it should even consider it all, which one might debate since the problem is really just that two users are trying to share the same cookie, which they obviously cannot given that cookies are specific to the browser and not tabs, meaning the user who made the last request will override the cookie of the other user (anonymous or not).

Using multiple tabs is not a problem under any other circumstance; the antiforgery token is validated against the security token stored in the cookie which is reused on all requests after it has initially been populated. See below:
https://github.com/dotnet/aspnetcore/blob/main/src/Antiforgery/src/Internal/DefaultAntiforgeryTokenGenerator.cs#L102
https://github.com/dotnet/aspnetcore/blob/main/src/Antiforgery/src/Internal/DefaultAntiforgery.cs#L278

It is only in the circumstance outlined by @blowdart's comment that it becomes an issue using multiple tabs, in which case, the problem isn't that you are using multiple tabs at all, but that you are trying to use multiple users in the same browser at the same time.

What lead me down this rabbit hole is that users of my application generally expect to use multiple tabs when working, whereby it would be a problem if antiforgery tokens wouldn't work in that scenario, but they do. 

I only ended up spending a couple hours on this, because the documentation stated otherwise.
